### PR TITLE
chore: API migration tests

### PR DESCRIPTION
**Pull request by awsluja**
#### Description of changes
This set of tests adds coverage for the following API sub categories:
- Basic REST API w/ dynamodb
- Lambda functions + Rest APIs
- Serverless REST APIs
- GraphQL v1 w/ functions
- GraphQL v2 w/ overrides

The approach for all of these tests is the same:
- Build up with v10
- Pull down with vLatest
- Compare CFN & validate assertions that should also be true

Passing E2E: https://app.circleci.com/pipelines/github/aws-amplify/amplify-cli/14448/workflows/290e683f-4062-41c4-aa5a-0d700397cea2

#### Checklist

- [x] PR description included
- [ ] `yarn test` passes
- [ ] Tests are [changed or added](https://github.com/aws-amplify/amplify-cli/blob/dev/CONTRIBUTING.md#tests)
- [ ] Relevant documentation is changed or added (and PR referenced)
- [ ] New AWS SDK calls or CloudFormation actions have been added to relevant test and service IAM policies
- [ ] [Pull request labels](https://github.com/aws-amplify/amplify-cli/blob/dev/CONTRIBUTING.md#labels) are added

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/amplify-migration-tests/src/__tests__/migration_tests_v10/api-graphql-v1.migration.test.ts [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+                permissions: ['storage'],
+                choices: ['api', 'function', 'storage'],
+                resources: ['Comment:@model(appsync)'],
+                resourceChoices: ['Post:@model(appsync)', 'Comment:@model(appsync)'],
+                operations: ['read'],
+                },
+            },
+            'nodejs',
+        );
+        await amplifyPushWithoutCodegen(projRoot, undefined, true);
```
Comment (sobolk) [inline packages/amplify-migration-tests/src/__tests__/migration_tests_v10/api-graphql-v1.migration.test.ts:63]: is pushing with latest codebase here expected ?

**Review by sobolk — COMMENTED** [review]
lgtm,
should some of this also go to category api repo ?

File: packages/amplify-migration-tests/src/__tests__/migration_tests_v10/api-graphql-v1.migration.test.ts (hunk 1 above)
Comment (awsluja) [reply packages/amplify-migration-tests/src/__tests__/migration_tests_v10/api-graphql-v1.migration.test.ts:63]: this method is weird; the middle parameter is for testingWithLatestCodebase

File: packages/amplify-migration-tests/src/__tests__/migration_tests_v10/__snapshots__/api-graphql-v1.migration.test.ts.snap [hunk 2]
```diff
@@ -0,0 +1,10 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
+
+exports[`api graphql v1 migration tests ...adds graphql with v1 transformer, updates models, adds function access, and pulls with latest version 1`] = `""`;
+
+exports[`api graphql v1 migration tests ...adds graphql with v1 transformer, updates models, adds function access, and pulls with latest version 2`] = `
+"Outputs
+[+] Output LambdaExecutionRoleArn: {\\"Value\\":{\\"Fn::GetAtt\\":[\\"LambdaExecutionRole\\",\\"Arn\\"]}}
```
Comment (awsluja) [inline packages/amplify-migration-tests/src/__tests__/migration_tests_v10/__snapshots__/api-graphql-v1.migration.test.ts.snap:7]: This LambdaExecutionRoleArn is new in 10.5.2 and forward, so this is good (we are expecting it to create this in vLatest)

The migration tests are setup with 10.5.1, which does not have this.

File: packages/amplify-migration-tests/src/__tests__/migration_tests_v10/__snapshots__/api-rest-basic.migration.test.ts.snap [hunk 3]
```diff
@@ -0,0 +1,10 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
+
+exports[`api REST migration tests ...adds rest APIs & DDB with v10 and pulls without drift in latest version 1`] = `""`;
+
+exports[`api REST migration tests ...adds rest APIs & DDB with v10 and pulls without drift in latest version 2`] = `
+"Outputs
+[+] Output LambdaExecutionRoleArn: {\\"Value\\":{\\"Fn::GetAtt\\":[\\"LambdaExecutionRole\\",\\"Arn\\"]}}
```
Comment (awsluja) [inline packages/amplify-migration-tests/src/__tests__/migration_tests_v10/__snapshots__/api-rest-basic.migration.test.ts.snap:7]: same thing here

File: packages/amplify-migration-tests/src/__tests__/migration_tests_v10/__snapshots__/api-rest-lambda.migration.test.ts.snap [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+"
+`;
+
+exports[`api lambda migration tests ...adds lambda with multiple rest APIs with v10 and pulls without drift in latest version 2`] = `
+"Outputs
+[+] Output LambdaExecutionRoleArn: {\\"Value\\":{\\"Fn::GetAtt\\":[\\"LambdaExecutionRole\\",\\"Arn\\"]}}
+
+Outputs
+[+] Output LambdaExecutionRoleArn: {\\"Value\\":{\\"Fn::GetAtt\\":[\\"LambdaExecutionRole\\",\\"Arn\\"]}}
```
Comment (awsluja) [inline packages/amplify-migration-tests/src/__tests__/migration_tests_v10/__snapshots__/api-rest-lambda.migration.test.ts.snap:18]: and same thing here

**Review by sobolk — APPROVED** [review]

**Review by sobolk — DISMISSED** [review]

File: packages/amplify-migration-tests/src/__tests__/migration_tests_v10/api-graphql-v1.migration.test.ts [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+        deleteProjectDir(projRoot);
+    });
+
+    // inspired by function_5.test.ts
+    it('...adds graphql with v1 transformer, updates models, adds function access, and pulls with latest version', async () => {
+        projectName = 'graphqlmigration';
+        projRoot = await createNewProjectDir(projectName);
+        
+        await initJSProjectWithProfileV10(projRoot, { name: projectName, disableAmplifyAppCreation: false });
+        await addApiWithoutSchema(projRoot, { transformerVersion: 1 });
```
[the remark this answers is not in the document]
Comment (awsluja) [reply packages/amplify-migration-tests/src/__tests__/migration_tests_v10/api-graphql-v1.migration.test.ts:36]: got it, I will just remove this then, it doesn't make sense to have a duplicate test for v1 transformer

File: packages/amplify-migration-tests/src/__tests__/migration_tests_v10/api-rest-lambda.migration.test.ts [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+        });
+        await addRestApi(projRoot, {
+            isFirstRestApi: false,
+            existingLambda: true,
+            restrictAccess: true,
+            allowGuestUsers: false,
+        });
+
+        // add more paths to and test policy slicing
+        for (let i = 0; i < 15; i++) {
```
[the remark this answers is not in the document]
Comment (awsluja) [reply packages/amplify-migration-tests/src/__tests__/migration_tests_v10/api-rest-lambda.migration.test.ts:40]: not valid in a for(... i++) loop; const only works in for...of loops
image

File: packages/amplify-migration-tests/src/migration-helpers-v10/cfn-diff-exclusions.ts [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+        || resource.Type === 'AWS::ApiGateway::GatewayResponse'
+        ){
+          delete resources[key];
+        }
+        if(resource.Type === 'AWS::AppSync::ApiKey' && resource.Properties){
+          delete resource.Properties.Expires;
+        }
+      }
+    }
+    if(currentCategory === 'api'){
```
[the remark this answers is not in the document]
Comment (awsluja) [reply packages/amplify-migration-tests/src/migration-helpers-v10/cfn-diff-exclusions.ts:30]: they both have same issue w/ API gateway being the main problem

**Review by akshbhu — DISMISSED** [review]

File: packages/amplify-migration-tests/src/__tests__/migration_tests_v10/api-graphql-v2.migration.test.ts [hunk 8]
```diff
@@ -0,0 +1,74 @@
+import { addApiWithoutSchema,
```
Comment (jhockett) [inline packages/amplify-migration-tests/src/__tests__/migration_tests_v10/api-graphql-v2.migration.test.ts:1]: Please run prettier on the PR. Indentation looks off and single quotes should be the default choice for string literals.

File: packages/amplify-migration-tests/src/__tests__/migration_tests_v10/api-graphql-v2.migration.test.ts [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+        try {
+            await amplifyPull(projRoot2, { emptyDir: true, appId }, true);
+            assertNoParameterChangesBetweenProjects(projRoot, projRoot2);
+            expect(collectCloudformationDiffBetweenProjects(projRoot, projRoot2, cfnDiffExclusions)).toMatchSnapshot();
+            await amplifyPushAuth(projRoot2, true);
+            assertNoParameterChangesBetweenProjects(projRoot, projRoot2);
+            expect(collectCloudformationDiffBetweenProjects(projRoot, projRoot2, cfnDiffExclusions)).toMatchSnapshot();
+
+            // check overridden config in cloud after pushing with vLatest
+            const overriddenAppsyncApiOverride = await getAppSyncApi(GraphQLAPIIdOutput, region);
```
Comment (jhockett) [inline packages/amplify-migration-tests/src/__tests__/migration_tests_v10/api-graphql-v2.migration.test.ts:65]: nit: name can be shortened

```ts
  const overriddenAppsyncApi = ...
```

**Review by jhockett — APPROVED** [review]
Left various nitpicks, but overall LGTM

**Review by akshbhu — APPROVED** [review]
Looks good

**Pull request merged**